File: 30-days-v1/day-16-search-insert-position.ts
// https://leetcode.com/problems/search-insert-position/

/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */

/**
Given a sorted array of distinct integers and a target value, return the index if the target is found. If not, return the index where it would be if it were inserted in order.

You must write an algorithm with O(log n) runtime complexity.

Ex1
Input: nums = [1,3,5,6], target = 5
Output: 2

Ex 2
Input: nums = [1,3,5,6], target = 2
Output: 1

**/
var searchInsert = function(nums, target) {
    let start = 0;
    let end = nums.length - 1;
    
    while (start <= end) {
        let mid = Math.floor(start + (end - start) / 2); // prevents overflow
        
        if(nums[mid] == target) {
            return mid;
        } else {
            if (target < nums[mid]) {
                end = mid - 1;
            } else {
                start = mid + 1; 
            }
           
        }
    }
    
    return start;
    
};

File: 2022/30-days-v1/day-16-search-insert-position.py
# // https://leetcode.com/problems/search-insert-position/

# /**
#  * Given a sorted array of distinct integers and a 
#  * target value, return the index if the target is found. 
#  * If not, return the index where it would be if it were inserted in order.
#  * 
#  * You must write an algorithm with O(log n) runtime complexity.
#  */

# /**
#  * example 1:
#  * 
#  * Input: nums = [1,3,5,6], target = 5
#  * Output: 2
#  */

class Solution:
    def searchInsert(self, nums: List[int], target: int) -> int:
        
        try:
            index = nums.index(target)
            return index
        except ValueError:
            # if beginning of the array
            if target < nums[0]:
                return 0;
            # if target is great than the last value of the array
            # then it is the length of the array
            if target > nums[len(nums) - 1]:
                return len(nums)
            for num in range(len(nums)):
                if nums[num] < target and nums[num + 1] > target:
                    return num + 1
            
        
        